proxy/src/http.rs

```diff
@@ -147,67 +128,91 @@ where
 /// Asserts presence of the owner and reject the request early if missing. Otherwise unpacks and
 /// passes down.
 #[must_use]
-pub fn with_owner_guard<R>(
-    api: Arc<Mutex<coco::PeerApi>>,
-    registry: Shared<R>,
-    store: Arc<RwLock<kv::Store>>,
-) -> BoxedFilter<(coco::User,)>
+fn with_owner_guard<R>(ctx: Ctx<R>) -> BoxedFilter<(coco::User,)>
 where
     R: registry::Client + 'static,
 {
     warp::any()
-        .and(with_peer(api))
-        .and(with_shared(registry))
-        .and(with_shared(store))
-        .and_then(
-            |api: Arc<Mutex<coco::PeerApi>>, registry: Shared<R>, store: Arc<RwLock<kv::Store>>| async move {
-                let session =
-                    crate::session::current(Arc::clone(&api), &*registry.read().await, &*store.read().await)
-                        .await
-                        .expect("unable to get current sesison");
-
-                if let Some(identity) = session.identity {
-                    let api = api.lock().await;
-                    let user = coco::get_user(&*api, &identity.urn).expect("unable to get coco user");
-                    let user = coco::verify_user(user).expect("unable to verify user");
-
-                    Ok(user)
-                } else {
-                    Err(Rejection::from(error::Routing::MissingOwner))
-                }
-            },
-        )
+        .and(with_context(ctx))
+        .and_then(|ctx: Ctx<R>| async move {
+            let ctx = ctx.read().await;
+            let session = crate::session::current(&ctx.peer_api, &ctx.registry, &ctx.store)
+                .await
+                .expect("unable to get current sesison");
+
+            if let Some(identity) = session.identity {
+                let user = ctx
+                    .peer_api
+                    .get_user(&identity.urn)
+                    .expect("unable to get coco user");
+                let user = coco::verify_user(user).expect("unable to verify user");
+
+                Ok(user)
+            } else {
+                Err(Rejection::from(error::Routing::MissingOwner))
+            }
+        })
         .boxed()
 }
 
-/// Thread-safe container for threadsafe pass-through to filters and handlers.
-pub type Shared<T> = Arc<RwLock<T>>;
+/// Container to pass down dependencies into HTTP filter chains.
+pub struct Context<R> {
+    /// [`coco::Api`] to operate on the local monorepo.
+    peer_api: coco::Api,
+    /// Storage to manage keys.
+    keystore: keystore::Keystorage,
```

Comment:
Since the context gets cloned around a bunch, I wonder how this plays together with the security measures that I mentioned before?

Reply:
`Keystorage` is fine. It's just a handle to the file system and unencrypts the key from disk when it needs the key. Again, the second part of the equation is using this https://github.com/radicle-dev/radicle-upstream/pull/686/files#r458688223